### ISSUE TITLE
Provide a flag to enable "lazy" loading of NDS data to improve memory efficiency

### ISSUE THREAD

**Issue by SeanDS** [gwpy/gwpy#1160]
For analysing data at high sample rates, downloading many hours worth of data requires huge amounts of memory and makes such a task only really suitable for clusters/servers. But when computing for instance a spectrogram, the data requirement for any given stride is much smaller - e.g. 20 seconds worth - yet gwpy still has to download the whole dataset before starting to compute it.

It would probably require a major bit of work, but would it be possible to create a "lazy" `TimeSeries` object (e.g. with a special flag) that can get data for a specific segment from the server only when asked it? That would allow spectrograms to be computed in a much more memory efficient way at the expense of extra NDS requests.

**Comment by SeanDS** [comment]
Lee also has an interesting, related project [here](https://git.ligo.org/lee-mcculler/filament) which uses `asyncio` to fetch and cache NDS data. Perhaps this could be merged into gwpy eventually.

**Comment by duncanmmacleod** [comment]
@SeanDS, I would be happy to review a PR that implements something like `Spectrogram.get` that does a lazy fetch+FFT over each chunk.

**Comment by SeanDS** [comment]
Would it be feasible to implement this on the `nds2.py` level, so that all existing methods that use NDS data can choose to use lazy or eager retrieval? Maybe as a special kind of `DictClass`?
